databricks_blueprints/download_file_from_dbfs.py

```diff
@@ -0,0 +1,134 @@
+import argparse
+import os
+import sys
+import base64
+import requests
+import re
+import shipyard_utils as shipyard
+try:
+    import helpers
+    import errors
+except BaseException:
+    from . import helpers
+    from . import errors
+
+ONE_MB = 1024*1024
+
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--access-token', dest='access_token', required=True)
+    parser.add_argument('--instance-id', dest='instance_id', required=True)
+    parser.add_argument('--source-file-name', dest='source_file_name', required=True)
+    parser.add_argument('--source-folder-name', dest='source_folder_name', required=False)
+    parser.add_argument('--dest-file-name', dest='dest_file_name', required=False)
+    parser.add_argument('--dest-folder-name', dest='dest_folder_name', required=False)
+    parser.add_argument('--source-file-name-match-type', dest='source_file_name_match_type',
+                        choices={'exact_match', 'regex_match'}, required=True)
+    args = parser.parse_args()
+    return args
+
+
+def download_file_from_dbfs(client, source_file_path, dest_file_path):
+    read_handle_endpoint = "/dbfs/read"
+    # check if filetype works with binary or not
+    if ('.csv' in source_file_path) or ('.txt' in source_file_path):
+        is_binary = False
+        write_mode = "w+"
+    else:
+        is_binary = True
+        write_mode = 'wb+'
+    with open(dest_file_path, write_mode) as file:
+        offset = 0
+        while True:
+            payload = {
+                "path": source_file_path,
+                "offset": offset, 
+                "length": ONE_MB # 1MB of data 
+            }
+            try:
+                read_response = client.get(read_handle_endpoint, params=payload)
+            except Exception:
+                print(f'Connection Error for {read_handle_endpoint}')
+                print(f"Data downloaded: {offset/1024}MB")
+                sys.exit(errors.EXIT_CODE_UNKNOWN_ERROR)
+
+            if read_response.status_code == requests.codes.ok:
+                response_data = read_response.json()
+                data = base64.b64decode(response_data['data'])
+                if is_binary:
+                    file.write(data)
+                else:
+                    file.write(data.decode('utf-8'))
+                # check if there is still more data left to read
+                if response_data['bytes_read'] == ONE_MB:
+                    offset += ONE_MB
+                else: # that was the last of the data in the file
+                    break
```

Comment:
Is there no other way to know when the download is done? 
It seems weird that you have to check the response size against the request size to determine whether or not to continue. 
What if the response is less than ONE_MB - the service returns less than requested? Is the service guaranteeing to return the exact number of bytes requested, and does that play nicely with the base64 encoding? 
Can you link to docs for this endpoint?

Reply:
Documentation for reads comes here: https://docs.databricks.com/dev-tools/api/latest/dbfs.html#read
Most of the download logic itself was derived from how file uploads are handled though which also deals with base64 data: https://docs.databricks.com/dev-tools/api/latest/examples.html#upload-a-big-file-into-dbfs

Reply:
Okay, so it looks like you can use the response size being less than the requested size to determine if you can stop reading.
Can you test / provide an example of what happens when the `offset` equals the file size, and `length` is some positive value?

I think we should update the logic to introduce a `length` or `requested_read_length` parameter, use that value in the params, and then compare the resulting length to that value. If the resulting read is less than the request, then stop reading. Otherwise, continue. But, this depends on my question / example above. [](edited-with-Axolo)

Reply:
If i'm getting your example: you're asking say the offset is some value like say 10024 and the file size itself is 10024? That means the amount of data downloaded is equivalent to the amount of data the file has so even if another call is performed the next call will just return 0 bytes (no more data left to download) and close the file stream. 

> I think we should update the logic to introduce a `length` or `requested_read_length` parameter, use that value in the params, and then compare the resulting length to that value. If the resulting read is less than the request, then stop reading. Otherwise, continue. But, this depends on my question / example above.
logic of the code already works somewhat similarly to what you're suggesting unless you're requesting for the ability to perform partial downloads?

Reply:
Yes, your example is correct. I'm not asking for partial downloads, just that the example case is accounted for and works correctly.